src/library_wasi.js

```diff
@@ -282,10 +274,30 @@ var WasiLibrary = {
       }
       num += len;
     }
-#endif // SYSCALLS_REQUIRE_FILESYSTEM
+  },
+#endif
+
+#if SYSCALLS_REQUIRE_FILESYSTEM
+  fd_write__deps: [
+    '$doWritev',
+#if WASM_WORKERS
+    '$fd_write_nofs',
+#endif
+  ],
+  fd_write: (fd, iov, iovcnt, pnum) => {
+#if WASM_WORKERS
+    if (ENVIRONMENT_IS_WASM_WORKER) {
+      return fd_write_nofs(fd, iov, iovcnt, pnum);
```

Comment:
@juj @kripken is this the behaviour we want?

If the application is setup to use the real filesystem, we would end up with two different filesystem behaviours, one in the wasm workers (which would see a dummy/fake filesystem just for stdout) and a different behaviour on the main thread and on the pthreads?

Reply:
An alternative to doing this this way would be to monkey-patch all the filesystem functions on wasm worker startup so it sees the dummy/fake ones instead of the fully-functional ones?

Reply:
I was a little confused about how fd_write could even be called at all from worker.. it looks like we were not correctly injecting the assertion: https://github.com/emscripten-core/emscripten/pull/20343

Reply:
> @juj @kripken is this the behaviour we want?
> 
> If the application is setup to use the real filesystem, we would end up with two different filesystem behaviours, one in the wasm workers (which would see a dummy/fake filesystem just for stdout) and a different behaviour on the main thread and on the pthreads?

According to the previous discussion, there is currently no available filesystem in the WASM workers, so I think the behavior in the WASM workers should be consistent with `-sFILESYSTEM=0` (no filesystem), and this part of the logic can be reused.

Reply:
I guess that make some sense, but it also make the code larger because we are including both versions of the FS code at the same time.

Another approach might be simply to make all FS syscalls fail on wasm workers (like most other syscalls do). Users could still call `emscripten_console_log` and `emscripten_console_logf` to do write to the console but the wouldn't be able to use normal stdio.  Given that wasm workers are already massively restricted in what they can do perhaps this is a reasonable restriction?

Reply:
> I guess that make some sense, but it also make the code larger because we are including both versions of the FS code at the same time.
> 
> Another approach might be simply to make all FS syscalls fail on wasm workers (like most other syscalls do). Users could still call `emscripten_console_log` and `emscripten_console_logf` to do write to the console but the wouldn't be able to use normal stdio. Given that wasm workers are already massively restricted in what they can do perhaps this is a reasonable restriction?

I considered this option, but due to the widespread use of normal stdio it would be a big breaking change to disable this behavior in the WASM worker.

Reply:
@juj WDTY?

Reply:
> I was a little confused about how fd_write could even be called at all from worker.. it looks like we were not correctly injecting the assertion: #20343

Yeah, I think this will be a problem. Once we land that PR to fix wasm workers to correctly assert on unsupported proxied functionality, then I think this patch to library_wasi.js will stop working.

The infrastructure of how `printf()` is tied to file handles is really awkward here. Ideally, we would definitely want to support `printf()` function in Wasm Workers, since its use is so common.

But we do want #20343 more, because it will be able to identify unsupported things better.

I kind of wish we could start the `printf()` function implementation with a 

```c++
void printf(...) {
  if (ENVIRONMENT_IS_WASM_WORKER || NO_FILESYSTEM) printf_no_filesystem(...);
  else printf_regular(...);
}
```
so that we'd route the implementation away from any filesystem in these types of scenarios.

@medns I think it might be best to splice this printf/stdio stuff out from this PR, and first land a PR that only tests Wasm Workers in Node.js without printf/filesystem parts? Would that make sense?

Reply:
Converting the tests from using `printf` to `emscripten_console_logf` would probably be a good idea (at least for now).

Reply:
> > I was a little confused about how fd_write could even be called at all from worker.. it looks like we were not correctly injecting the assertion: #20343
> 
> Yeah, I think this will be a problem. Once we land that PR to fix wasm workers to correctly assert on unsupported proxied functionality, then I think this patch to library_wasi.js will stop working.
> 
> The infrastructure of how `printf()` is tied to file handles is really awkward here. Ideally, we would definitely want to support `printf()` function in Wasm Workers, since its use is so common.
> 
> But we do want #20343 more, because it will be able to identify unsupported things better.
> 
> I kind of wish we could start the `printf()` function implementation with a
> 
> ```c++
> void printf(...) {
>   if (ENVIRONMENT_IS_WASM_WORKER || NO_FILESYSTEM) printf_no_filesystem(...);
>   else printf_regular(...);
> }
> ```
> 
> so that we'd route the implementation away from any filesystem in these types of scenarios.
> 
> @medns I think it might be best to splice this printf/stdio stuff out from this PR, and first land a PR that only tests Wasm Workers in Node.js without printf/filesystem parts? Would that make sense?

> > I was a little confused about how fd_write could even be called at all from worker.. it looks like we were not correctly injecting the assertion: #20343
> 
> Yeah, I think this will be a problem. Once we land that PR to fix wasm workers to correctly assert on unsupported proxied functionality, then I think this patch to library_wasi.js will stop working.
> 
> The infrastructure of how `printf()` is tied to file handles is really awkward here. Ideally, we would definitely want to support `printf()` function in Wasm Workers, since its use is so common.
> 
> But we do want #20343 more, because it will be able to identify unsupported things better.
> 
> I kind of wish we could start the `printf()` function implementation with a
> 
> ```c++
> void printf(...) {
>   if (ENVIRONMENT_IS_WASM_WORKER || NO_FILESYSTEM) printf_no_filesystem(...);
>   else printf_regular(...);
> }
> ```
> 
> so that we'd route the implementation away from any filesystem in these types of scenarios.
> 
> @medns I think it might be best to splice this printf/stdio stuff out from this PR, and first land a PR that only tests Wasm Workers in Node.js without printf/filesystem parts? Would that make sense?

Okay, let me strip out the modifications about filesystem and use `emscripten_console_logf` instead of `printf` to implement unit tests.
I'm on vacation for the next few days and will submit new patch in early October.

Reply:
Modified as required.